crates/da-framework/src/queue.rs

```diff
@@ -0,0 +1,231 @@
+    let is_new_entries = (v >> TAIL_BIT_SHIFT) > 0;
+    (is_new_entries, incr)
+}
+
+/// Encodes the "head word".
+fn encode_head(new_entries: bool, v: IncrTy) -> HeadTy {
+    if v > HEAD_WORD_INCR_MASK {
+        panic!("da/queue: tried to increment front by too much {v}");
+    }
+
+    ((new_entries as IncrTy) << TAIL_BIT_SHIFT) | (v & HEAD_WORD_INCR_MASK)
+}
+
+/// Builder for [`DaQueue`].
+pub struct DaQueueBuilder<Q: DaQueueTarget> {
+    // FIXME use `IdxTy`
+    original_front_pos: IncrTy,
+    new_front_pos: IncrTy,
+    original_next_pos: IncrTy,
+    new_entries: Vec<Q::Entry>,
+}
+
+impl<Q: DaQueueTarget> DaQueueBuilder<Q> {
+    /// Returns what would be the idx of the next element to be added to the
+    /// queue.
+    pub fn next_idx(&self) -> usize {
+        self.original_next_pos as usize + self.new_entries.len()
+    }
+
+    /// Tries to add to the increment to the front of the queue.
+    ///
+    /// Returns if successful, fails if overflow.
+    pub fn add_front_incr(&mut self, incr: IncrTy) -> bool {
+        // TODO add checks to only allow increments if there's entries
+        let incr_front = self.new_front_pos - self.original_front_pos;
+        let new_front = (self.new_front_pos as u64) + (incr as u64);
+
+        // So we don't overrun the back of the entries that'd be added.
+        if new_front >= self.next_idx() as u64 {
+            return false;
+        }
+
+        let new_incr = (incr as u64) + (incr_front as u64);
+        if new_incr >= HEAD_WORD_INCR_MASK as u64 {
+            false
+        } else {
+            self.new_front_pos = new_front as IncrTy;
+            true
+        }
+    }
+
+    /// Appends an entry to the queue.
+    pub fn append_entry(&mut self, e: Q::Entry) {
+        // FIXME this doesn't account for bounds checks
+        self.new_entries.push(e);
+    }
+
+    /// Returns the count of new entries we added that would be consumed.
+    fn consumed_new_entries(&self) -> usize {
+        // FIXME this doesn't account for bounds checks
+        let new_front = self.new_front_pos as i64;
+        let orig_next = self.original_next_pos as i64;
+        let overrun = new_front - orig_next;
+        (overrun as usize).clamp(0, usize::MAX)
+    }
+}
+
+impl<Q: DaQueueTarget> DaBuilder<Q> for DaQueueBuilder<Q> {
+    type Write = DaQueue<Q>;
+
+    fn from_source(t: Q) -> Self {
+        Self {
+            original_front_pos: t.cur_front(),
+            new_front_pos: t.cur_front(),
+            original_next_pos: t.cur_next(),
+            new_entries: Vec::new(),
+        }
+    }
+
+    fn into_write(mut self) -> Result<Self::Write, BuilderError> {
+        // Remove things from this that are redundant.
+        self.new_entries.drain(..self.consumed_new_entries());
+
+        let tail = self.new_entries;
+        let incr_front = self.new_front_pos - self.original_front_pos;
+        Ok(DaQueue { tail, incr_front })
+    }
+}
+
+// TODO tests for the wacky logic up ^there
```

Comment:
Could you add these? If it is not feasible right now, please create ticket(s). Ideally we want every module to have tests which would also greatly help on reviewing and understanding the code.